Immutability error for function parameters caught at runtime instead of compile-time

## Summary

When attempting to assign to an immutable function parameter (struct), the error is caught at **runtime** instead of **compile-time**. This should be a typechecker error.

## Reproduction

```ez
do game_loop(hero lib.Hero) {
    // ... later in code ...
    hero = updatedHero  // This should fail at compile time
}
```

## Expected Behavior

The typechecker should catch this during compilation and produce an error like:
```
error[E5013]: cannot assign to immutable variable 'hero' (declared as const)
  --> file.ez:XX:XX
```

## Actual Behavior

The code compiles successfully, and the error only appears at **runtime** when the assignment is executed:

```
error[E5013]: cannot assign to immutable variable 'hero' (declared as const)
  --> EZ-TAG/main/main.ez:635:17
   |
635 |                 hero = updatedHero
   |                 ^ error occurred here
```

## Why This Matters

- Users expect compile-time errors for static analysis issues
- Runtime errors for immutability violations can cause unexpected crashes in production
- This makes debugging harder as the error only surfaces when that code path is executed

## Workaround

Use `copy()` at the start of the function:
```ez
do game_loop(heroIn lib.Hero) {
    temp hero = copy(heroIn)
    // Now hero is mutable
}
```

## Affected Code

Likely in `pkg/typechecker/typechecker.go` - the assignment check for function parameters should verify mutability at compile time.

--- a/cmd/ez/main.go
+++ b/cmd/ez/main.go
@@ -116,7 +116,15 @@ func printVersion() {
 }
 
 func checkFile(filename string) {
-	data, err := os.ReadFile(filename)
+	// Resolve to absolute path
+	absFile, err := filepath.Abs(filename)
+	if err != nil {
+		fmt.Printf("Error resolving path: %v\n", err)
+		return
+	}
+	absDir := filepath.Dir(absFile)
+
+	data, err := os.ReadFile(absFile)
 	if err != nil {
 		fmt.Printf("Error reading file: %v\n", err)
 		return
@@ -124,7 +132,7 @@ func checkFile(filename string) {
 
 	source := string(data)
 	l := lexer.NewLexer(source)
-	p := parser.NewWithSource(l, source, filename)
+	p := parser.NewWithSource(l, source, absFile)
 	program := p.ParseProgram()
 
 	// Check for lexer errors
@@ -139,7 +147,7 @@ func checkFile(filename string) {
 				code = errors.ErrorCode{Code: lexErr.Code, Name: "lexer-error", Description: "Lexer error"}
 			}
 			sourceLine := errors.GetSourceLine(source, lexErr.Line)
-			ezErr := errors.NewErrorWithSource(code, lexErr.Message, filename, lexErr.Line, lexErr.Column, sourceLine)
+			ezErr := errors.NewErrorWithSource(code, lexErr.Message, absFile, lexErr.Line, lexErr.Column, sourceLine)
 			errList.AddError(ezErr)
 		}
 		fmt.Print(errors.FormatErrorList(errList))
@@ -157,8 +165,8 @@ func checkFile(filename string) {
 		fmt.Print(errors.FormatErrorList(p.EZErrors()))
 	}
 
-	// Type checking
-	tc := typechecker.NewTypeChecker(source, filename)
+	// Type checking for main file
+	tc := typechecker.NewTypeChecker(source, absFile)
 	tc.CheckProgram(program)
 
 	// Check for type errors
@@ -170,6 +178,112 @@ func checkFile(filename string) {
 	// Display type checker warnings
 	if tc.Errors().HasWarnings() {
 		fmt.Print(errors.FormatErrorList(tc.Errors()))
+	}
+
+	// Initialize the module loader to check imported modules
+	loader := interpreter.NewModuleLoader(absDir)
+
+	// Track all modules that need to be checked
+	filesChecked := 1 // Start with 1 for the main file
+	hasErrors := false
+	modulesToCheck := collectImports(program, absDir, absFile)
+	checked := make(map[string]bool)
+	checked[absFile] = true
+
+	for len(modulesToCheck) > 0 {
+		// Pop first module
+		modPath := modulesToCheck[0]
+		modulesToCheck = modulesToCheck[1:]
+
+		if checked[modPath] {
+			continue
+		}
+		checked[modPath] = true
+
+		// Load the module using the loader
+		loader.SetCurrentFile(absFile) // Set context for relative imports
+		mod, err := loader.Load(modPath)
+		if err != nil {
+			fmt.Printf("Error loading module %s: %v\n", modPath, err)
+			hasErrors = true
+			continue
+		}
+
+		// Check each file in the module
+		for _, filePath := range mod.Files {
+			if checked[filePath] {
+				continue
+			}
+			checked[filePath] = true
+			filesChecked++
+
+			// Read and check the file
+			fileData, err := os.ReadFile(filePath)
+			if err != nil {
+				fmt.Printf("Error reading %s: %v\n", filePath, err)
+				hasErrors = true
+				continue
+			}
+
+			fileSource := string(fileData)
+			fileLex := lexer.NewLexer(fileSource)
+			fileParser := parser.NewWithSource(fileLex, fileSource, filePath)
+			fileProgram := fileParser.ParseProgram()
+
+			// Check for lexer errors
+			if len(fileLex.Errors()) > 0 {
+				errList := errors.NewErrorList()
+				for _, lexErr := range fileLex.Errors() {
+					var code errors.ErrorCode
+					switch lexErr.Code {
+					case "E1005":
+						code = errors.E1005
+					default:
+						code = errors.ErrorCode{Code: lexErr.Code, Name: "lexer-error", Description: "Lexer error"}
+					}
+					sourceLine := errors.GetSourceLine(fileSource, lexErr.Line)
+					ezErr := errors.NewErrorWithSource(code, lexErr.Message, filePath, lexErr.Line, lexErr.Column, sourceLine)
+					errList.AddError(ezErr)
+				}
+				fmt.Print(errors.FormatErrorList(errList))
+				hasErrors = true
+				continue
+			}
+
+			// Check for parser errors
+			if fileParser.EZErrors().HasErrors() {
+				fmt.Print(errors.FormatErrorList(fileParser.EZErrors()))
+				hasErrors = true
+				continue
+			}
+
+			if fileParser.EZErrors().HasWarnings() {
+				fmt.Print(errors.FormatErrorList(fileParser.EZErrors()))
+			}
+
+			// Type check (skip main() requirement for module files)
+			fileTc := typechecker.NewTypeChecker(fileSource, filePath)
+			fileTc.SetSkipMainCheck(true) // Module files don't need main()
+			fileTc.CheckProgram(fileProgram)
+
+			if fileTc.Errors().HasErrors() {
+				fmt.Print(errors.FormatErrorList(fileTc.Errors()))
+				hasErrors = true
+				continue
+			}
+
+			if fileTc.Errors().HasWarnings() {
+				fmt.Print(errors.FormatErrorList(fileTc.Errors()))
+			}
+
+			// Collect more imports from this file
+			newImports := collectImports(fileProgram, absDir, filePath)
+			modulesToCheck = append(modulesToCheck, newImports...)
+		}
+	}
+
+	if hasErrors {
+		return
 	}
 
 	fmt.Printf("Check successful: %s\n", filename)

--- a/integration-tests/fail/multi-file/E5016_immutable_param_in_module/main.ez
+++ b/integration-tests/fail/multi-file/E5016_immutable_param_in_module/main.ez
@@ -1,0 +1,15 @@
+/*
+ * Error Test: E5016 - immutable-parameter in module file
+ * Expected: "cannot modify" (error in imported module should be caught)
+ * Regression test for bug #535
+ */
+
+import @std
+import lib"./mylib"
+
+using std
+
+do main() {
+    temp h = lib.Hero{name: "Test", hp: 50}
+    lib.update_hero(h)
+}

--- a/integration-tests/fail/multi-file/E5016_immutable_param_in_module/mylib/lib.ez
+++ b/integration-tests/fail/multi-file/E5016_immutable_param_in_module/mylib/lib.ez
@@ -1,0 +1,10 @@
+module mylib
+
+const Hero struct {
+    name string
+    hp int
+}
+
+do update_hero(hero Hero) {
+    hero = Hero{name: "Modified", hp: 999}
+}

